### DISCUSSION TITLE
Is predefined fields more optimized than dynamic fields?

### DISCUSSION THREAD

**Discussion by ducanh997** [milvus-io/milvus#36247] [Q&A and General discussion]
Hi,
I'm trying to understand the performance differences between predefined fields and dynamic fields.

From what I've gathered, predefined fields are declared in the schema when the collection is created, while dynamic fields can be added during data insertion without being explicitly defined in the schema.

My main concern is about performance:
Are predefined fields more optimized for querying compared to dynamic fields?
Which one is generally preferred in a production setting, especially when handling large datasets for vector search?
Any insights or experiences with Milvus in this regard would be helpful. Thanks!

**Comment by yhmo** [comment]
A predefined field is stored as column-based files. The dynamic field is stored as a list of JSON strings. It is easier to read/parse column-based file than JSON strings.

Currently, Milvus supports scalar index for the predefined fields, but doesn't support index for dynamic field.

In practice, if you want to do filtering-search on a property, you can define the property as a scalar field so that the filtering-search can use scalar index to improve search performance.
